File: 2025/zig/src/day04.zig
const std = @import("std");
const adlib = @import("adlib.zig");

fn accessibleRolls(grid: [][]u8, remove: bool) u32 {
    var accessible: u32 = 0;

    for (grid, 0..) |row, i| {
        for (row, 0..) |byte, j| {
            if (byte == '.') continue;

            var blocked: u8 = 0;
            if (j > 0 and row[j - 1] != '.') {
                blocked += 1;
            }
            if (j < row.len - 1 and row[j + 1] != '.') {
                blocked += 1;
            }
            if (i > 0 and grid[i - 1][j] != '.') {
                blocked += 1;
            }
            if (i < grid.len - 1 and grid[i + 1][j] != '.') {
                blocked += 1;
            }

            if (j > 0 and i > 0 and grid[i - 1][j - 1] != '.') {
                blocked += 1;
            }
            if (j > 0 and i < grid.len - 1 and grid[i + 1][j - 1] != '.') {
                blocked += 1;
            }

            if (j < row.len - 1 and i > 0 and grid[i - 1][j + 1] != '.') {
                blocked += 1;
            }
            if (j < row.len - 1 and i < grid.len - 1 and grid[i + 1][j + 1] != '.') {
                blocked += 1;
            }

            if (blocked <= 3) {
                accessible += 1;
                grid[i][j] = 'x';
            }
        }
    }

    if (remove) {
        for (grid) |row| {
            for (row) |*byte| {
                if (byte.* == 'x') byte.* = '.';
            }
        }
    }

    return accessible;
}

fn partOne(grid: [][]u8) !u32 {
    return accessibleRolls(grid, false);
}

fn partTwo(grid: [][]u8) !u32 {
    var accessible: u32 = 0;
    while (true) {
        const newly_accessible = accessibleRolls(grid, true);
        if (newly_accessible == 0) break;
        accessible += newly_accessible;
    }
    return accessible;
}

pub fn main() !void {
    var buf: [4096]u8 = undefined;
    const input = try adlib.inputFile("4");
    var reader = input.reader(&buf);
    const grid = try adlib.streamGrid(adlib.allocator, &reader.interface);
    const res_1 = try partOne(grid);
    try reader.seekTo(0);
    const res_2 = try partTwo(grid);
    std.debug.print("part one: {d}\npart two {d}\n", .{ res_1, res_2 });
    adlib.freeGrid(adlib.allocator, grid);
    input.close();
}

File: 2025/ts/src/day04.ts
import * as adlib from "./adlib.ts";

function boxAt(grid: string[][], row: number, col: number): 1|0 {
    return grid[row][col] === "." ? 0 : 1;
}

function countAccessibleBoxes(grid: string[][]): number {
    let accessible_boxes = 0;
    for (let i = 0; i < grid.length; i++) {
        const line = grid[i];
        for (let j = 0; j < line.length; j++) {
            if (!boxAt(grid, i, j)) continue;
            let box_count = 0;
            const space_left = j > 0;
            const space_right = j < line.length - 1;
            const space_up = i > 0;
            const space_down = i < grid.length - 1;

            if (space_left) {
                box_count += boxAt(grid, i, j - 1);
                if (space_down) box_count += boxAt(grid, i + 1, j - 1);
                if (space_up) box_count += boxAt(grid, i - 1, j - 1);
            }
            if (space_right) {
                box_count += boxAt(grid, i, j + 1)
                if (space_down) box_count += boxAt(grid, i + 1, j + 1);
                if (space_up) box_count += boxAt(grid, i - 1, j + 1);
            }
            if (space_down) box_count += boxAt(grid, i + 1, j);
            if (space_up) box_count += boxAt(grid, i - 1, j);
            if (box_count < 4) {
                accessible_boxes += 1;
                grid[i][j] = "x";
            }

        }
    }

    return accessible_boxes;
}

function partOne(input: string): number {
    const grid = input.trim().split("\n").map((s) => s.split(""));
    return countAccessibleBoxes(grid);
}

function partTwo(input: string): number {
    const grid = input.trim().split("\n").map((s) => s.split(""));
    let total_accessible = 0;
    while (true) {
        const newly_accessible = countAccessibleBoxes(grid);
        if (newly_accessible == 0) break;
        total_accessible += newly_accessible;

        for (let i = 0; i < grid.length; i++) {
            const row = grid[i];
            for (let j = 0; j < row.length; j++) {
                if (row[j] === "x") row[j] = ".";
            }
        }
    }
    return total_accessible;
}

export async function main() {
    const input = await new Response(Deno.stdin.readable).text();

    const res = partOne(input);
    const res2 = partTwo(input);

    console.log(`part one: ${res}\npart two: ${res2}\n`);
}

if (import.meta.main) await main();
